### PR TITLE
Support udp trackers

### PR DESCRIPTION
**Pull request by matei-oltean**

### CODE REVIEW

File: client/tracker.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+	binary.BigEndian.PutUint32(request[12:], transactionID)
+	copy(request[16:], torrentFile.Hash[:])
+	copy(request[36:], clientID[:])
+	binary.BigEndian.PutUint64(request[56:], downloaded)
+	binary.BigEndian.PutUint64(request[64:], left)
+	binary.BigEndian.PutUint64(request[72:], uploaded)
+	binary.BigEndian.PutUint32(request[80:], event)
+	binary.BigEndian.PutUint32(request[84:], ipAddr)
+	binary.BigEndian.PutUint32(request[88:], key)
+	binary.BigEndian.PutUint32(request[92:], uint32(expectedPeers))
```
Comment (matei-oltean) [inline client/tracker.go]: TODO: request -1 and read with timeout instead

File: client/tracker.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+	aAnnounce
+	aScrap
+	aError
+)
+
+// expectedPeers is the number of peers we are requesting
+const expectedPeers int = 20
+
+// getPeers returns the list of peers from a torrent file and client ID
+func getPeers(torrentFile *fileutils.TorrentFile, clientID [20]byte) ([]string, error) {
```
Comment (matei-oltean) [inline client/tracker.go]: Move it to fileUtils and make change it to `func (t *TorrentFile) getPeers ...` instead?

File: client/tracker.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+	if err != nil {
+		return nil, err
+	}
+	switch url.Scheme {
+	case "http", "https":
+		return getPeersHTTPS(torrentFile, clientID)
+	case "udp", "udp4":
+		return getPeersUDP(torrentFile, clientID, url, false)
+	case "udp6":
+		return getPeersUDP(torrentFile, clientID, url, true)
```
Comment (mookid) [inline client/tracker.go]: I would keep ` url.Scheme ` here

File: client/tracker.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+	// since we are using udp, we retry
+	// 8 times, with an increasing deadline
+	for try := 0; try < 8; try++ {
+		conn.SetDeadline(time.Now().Add(15 * (1 << try) * time.Second))
+		connID, err := connectToUDP(conn)
+		if err != nil {
```
Comment (matei-oltean) [inline client/tracker.go]: Wooops this is wrong, should check for timeout before.

File: client/tracker.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+	// uint32 interval
+	// uint32 leechers
+	// uint32 seeders
+	// list of addresses: uint32 IPv4/16 byte IPv6 address + 16 bit port
+	response := make([]byte, 20+6*expectedPeers)
+	_, err = conn.Read(response)
+	if err != nil {
+		return nil, err
+	}
+	if len(response) < 20 {
```
Comment (matei-oltean) [inline client/tracker.go]: `len(response)` is always >= 20 :|

**Pull request merged**